### ISSUE TITLE
Child process dont finished when parent close

### ISSUE THREAD

**Issue by oleksis** [Nuitka/Nuitka#970]
* Nuitka version, full Python version and Platform (Windows, OSX, Linux ...)

python -m nuitka --version
0.6.11.2
Python: 3.8.6 (default, Oct  1 2020, 13:01:33)  [GCC 10.2.0 64 bit (AMD64)]
Executable: D:/PYTHON/picta-dl/env/bin/python.exe
OS: Windows
Arch: x86_64

* How did you install Nuitka and Python

Using virtualenv 

* Also supply a [Short, Self Contained, Correct, Example](http://sscce.org/)
```python
#!/usr/bin/env python
# child.py
#   Compile with Nuitka:
#   python -m nuitka --onefile --mingw64 --assume-yes-for-downloads --windows-company-name=MyCompany --windows-product-name=child --windows-product-version=1.0.0.0 --windows-file-description="Child" child.py 


import time


def main_ok():
    try:
        while True:
            print('Im working...')
            time.sleep(2)
    except KeyboardInterrupt:  
        print('Exit from child!')

def main():
    # Dont work with CTRL_C_EVENT ??!
    while True:
        print('Im working...')
        time.sleep(2)


if __name__ == '__main__':
    main()
```

* Freeze

python -m pip freeze
Nuitka==0.6.11.2

**Comment by oleksis** [comment]
I try too:
pip install -U "https://github.com/Nuitka/Nuitka/archive/0.6.11.5.zip"

* Freeze
python -m pip freeze
Nuitka==0.6.11.5

The same behavior. Any advice?

**Comment by kayhayen** [comment]
I fixed this the other day, right now I am adding tests. I believe this to be a Windows only issue, because its not using 3rd party tech there, that solves all the things: https://nuitka.net/doc/factory.html

**Comment by kayhayen** [comment]
In my manual testing, CTRL-C on a sleeping process gave it `KeyboardInterrupt` exception, and that's what I am adding automated tests for now, also to prove my Linux ideas. I am not entirely sure what CTRL-Break is, but since the code also handles logoff and system restart, I think GUI applications without console out to also working.

**Comment by kayhayen** [comment]
So in my testing, this works when you CTRL-C from the terminal. CPython only implements sending CTRL-C to processes in the same group, otherwise it's hard kills. From task manager, killed processes get hard killed, only apps get soft killed it seems, but you have to appear there first. That seems to be as good as you can get in an obvious way.

What we could attempt is a monitoring (C only) thread in the Python part that will detect if the parent process is gone, and it if it is, to goes to a Python and injects the `KeyboardInterupt` manually, and potentially even ensures the file system cleanup, although deleting running code is not going to come easy, but it can be scheduled for the next boot I guess. 

Adding that kind of thread should be easy. For Windows Services in Nuitka Commercial, we already have similar stuff, I can take it from there.

**Comment by oleksis** [comment]
> What we could attempt is a monitoring (C only) thread in the Python part that will detect if the parent process is gone, and it if it is, to goes to a Python and injects the `KeyboardInterupt` manually, and potentially even ensures the file system cleanup, although deleting running code is not going to come easy, but it can be scheduled for the next boot I guess.

Here a more elaborate idea.  Imagine that the parent process also behaves like a GUI when closing the application.  The child process would stay until it terminates or receives the `KeyboardInterupt` and the child process should handle `KeyboardInterupt`.  

```python
#!/usr/bin/env python
# main.py
#   Compile with Nuitka:
#   python -m nuitka --onefile --mingw64 --assume-yes-for-downloads --windows-company-name=MyCompany --windows-product-name=main --windows-product-version=1.0.0.0 --windows-file-description="Main" main.py 

import os, sys, subprocess,time


def main():
    print('Starting child process ...')
    # child_proc = subprocess.Popen('python child.py', stdout=sys.stdout, stderr=sys.stderr)  # OK!!
    
    kwargs = {}
    
    if os.name == 'nt':
        kwargs['creationflags'] = subprocess.CREATE_NEW_PROCESS_GROUP
    # Running code...
    child_proc = subprocess.Popen('child.exe',
                                  stdout=sys.stdout,
                                  stderr=subprocess.STDOUT,
                                  **kwargs)  # OK?! Dont work with CTRL_C_EVENT ??!
    
    print('Waiting for kill child process in 5 seconds ...')
    print('Try CTRL + C ?')
    time.sleep(5)
    
    print('Killing child process')
    child_proc.kill()
    print("Exit from main!")


if __name__ == '__main__':
    main()

```

Thanks for your answers and I will be aware of the solution

**Comment by oleksis** [comment]
Now i try the factory and get:

```bash
pip install -U "https://github.com/Nuitka/Nuitka/archive/factory.zip"
```

```bash
python -m nuitka --onefile --mingw64 --assume-yes-for-downloads --windows-company-name=MyCompany --windows-product-name=child --windows-product-version=1.0.0.0 --windows-file-description="Child" child.py
```

```bash
./child.exe
Error, failed to created payload path.
No se puede crear un archivo que ya existe.
```

Any idea?

**Comment by kayhayen** [comment]
I guess, that's not good, in that case, it should tolerate it already existing, if you run it twice e.g. I will fix that too. Still struggling with error checking it seems.

**Comment by kayhayen** [comment]
ok, please retry, factory has  the suicide with KeyboardInterupt sent into itself when it notices that the launching program is exited.

One minor thing is that I would also add a unconditional kill e.g. 5 seconds after the `KeyboardInterrupt` was initiated. That will make sure, we follow unconditional kills of the user no matter what. But that's not there yet, should be easy though. Right now the monitoring thread simply exits after triggering the `KeyboardInterrupt` in the Python parts, but it could sleep another 5 seconds and then trigger `TerminateProcess` Win32 API, so that graceful exit has had its chance, but if that fails, too bad.

Also, unfortunately `KeyboardInterrupt` in Nuitka is only delivered later anyway, even now, `CTRL-C` will happen not after every bodycode execution, but possibly only the next time something happens. So testing is a bit imprecise necessarily.

Under Linux it worked all the time, only the Windows was affected indeed.

**Comment by oleksis** [comment]
> ok, please retry, factory has the suicide with KeyboardInterupt sent into itself when it notices that the launching program is exited.

* From Factory
```bash
pip install -U "https://github.com/Nuitka/Nuitka/archive/factory.zip"
```

* Freeze
```bash
pip freeze
Nuitka==0.6.12rc5
```

* Test and Output
```bash
(env)
oleksis@PC MINGW64 /d/python/test
$ ./child.exe
Im working...
Im working...
Im working...

(env)
oleksis@PC MINGW64 /d/python/test
$ Im working...
Traceback (most recent call last):
  File "C:\Users\oleksis\AppData\Local\MyCompany\child\1.0.0.0\child.py", line 33, in <module>
  File "C:\Users\oleksis\AppData\Local\MyCompany\child\1.0.0.0\child.py", line 28, in main
KeyboardInterrupt

```

Im glad look like is working now 🎉

**Other question...**

How can turn off the Traceback  from the child? Any way to say from Python or Compile time?

Thanks one more time. Good work with Nuitka 🥳

**Comment by oleksis** [comment]
Old school

```bash
(env)
oleksis@PC MINGW64 /d/python/test
$ ./child.exe 2> /dev/null
Im working...
Im working...
Im working...
Im working...
```

But i will like turn off the traceback in compile time ?!

**Comment by kayhayen** [comment]
Nuitka aims at full compatibility, i.e. it will give precisely the same stack traces as normal Python does. Actually e.g. including source code in the standalone distribution is something that I want to implement.

If you want to hide information, the Nuitka commercial plugins will cover that, and there is currently early bird work for these going on. They will e.g. also allow encrypted outputs of tracebacks, so you could decode them, but not your clients. If you yourself, e.g. not willing to buy those, can catch and swallow all exceptions, that will of course often also work.

**Comment by kayhayen** [comment]
If you consider going the commercial route, visit this https://nuitka.net/pages/commercial.html

**Comment by kayhayen** [comment]
Part of the current pre-release, I expect to make a release relatively soon.

**Comment by oleksis** [comment]
> Part of the current pre-release, I expect to make a release relatively soon.

Very happy to help or report this issue.  When you think it is convenient to close the Issue  👍🏾

**Comment by kayhayen** [comment]
Released as part of 0.6.12
